episode_sequence_generation.py

```diff
@@ -0,0 +1,269 @@
+    team_times = []
+
+    print('---------------- TEAMS -------------------------')
+
+    mcats = list(micro.keys())
+    for tid, team in team_data.items():
+        print(tid, sep=' ', end=' ', flush=True)
+        team_host_episodes = dict()
+        _team_times = dict()
+        for attacker_victim, alerts in team:
+            if len(alerts) <= 1:
+                continue
+
+            # Alert format: (dst_ip, mcat, ts, dst_port, signature)
+            # print(attacker_victim, len([(x[1]) for x in alerts])) # TODO: what about IPs not related to attacker?
+            first_elapsed_time = round((alerts[0][2] - start_times[tid]).total_seconds(), 2)
+
+            _team_times['->'.join(attacker_victim)] = first_elapsed_time
+
+            ts = [x[2] for x in alerts]
+            diff_ts = [0.0]
+            for i in range(1, len(ts)):
+                diff_ts.append(round((ts[i] - ts[i - 1]).total_seconds(), 2))
+            elapsed_time = list(accumulate(diff_ts))
+            relative_elapsed_time = [round(x + first_elapsed_time, 2) for x in elapsed_time]
+
+            host_episodes = []
+            for mcat in mcats:
+                # 2.5-minute (150s) fixed step (window). Can be reduced or increased depending on required granularity
+                hyperalert_seq = []
+                for i in range(int(first_elapsed_time), int(relative_elapsed_time[-1]), step):
+                    window = [a for dt, a in zip(relative_elapsed_time, alerts) if (i <= dt < (i + step)) and a[1] == mcat]
+                    hyperalert_seq.append(window)  # Alerts per 'step' seconds (window)
+
+                raw_episodes = _get_episodes(hyperalert_seq, micro[mcat], plot=plot)
+                if len(raw_episodes) > 0:
+                    for epi in raw_episodes:
+                        hyperalert_seq_epi = hyperalert_seq[epi[0]:epi[1]+1]
+                        episode = _create_episode(hyperalert_seq_epi, mcat, start_times[tid])
+                        host_episodes.append(episode)
+
+            if len(host_episodes) == 0:
+                continue
+
+            host_episodes.sort(key=lambda tup: tup[0])
+            team_host_episodes[attacker_victim] = host_episodes
+
+            if plot_alert_volumes:
+                plot_alert_volume_per_episode(tid, attacker_victim, host_episodes, mcats)
+
+        team_episodes.append(team_host_episodes)
+        team_times.append(_team_times)
+    return team_episodes, team_times
+
+
+# Step 3: Create episode sequences
+# Host = [connections] instead of team level representation
+def host_episode_sequences(team_episodes):
+    host_data = {}
+    print('# teams:', len(team_episodes))
+    print('----- TEAMS -----')
+    for tid, team in enumerate(team_episodes):
+        print(tid, sep=' ', end=' ', flush=True)
+        for (attacker, victim), episodes in team.items():
+            if len(episodes) < 2:
+                continue
+            # if ('10.0.0' in attacker or '10.0.1' in attacker):
```

Comment:
Can this part be deleted? It has never been uncommented (neither for CPTC nor for CCDC), so I assume it is outdated

Reply:
no, it's for testing